cpp/src/parquet/arrow/arrow_reader_writer_test.cc

```diff
@@ -3199,6 +3200,36 @@ TEST(TestArrowReaderAdHoc, HandleDictPageOffsetZero) {
   TryReadDataFile(test::get_data_file("dict-page-offset-zero.parquet"));
 }
 
+TEST(TestArrowReaderAdHoc, WriteBatchedNestedNullableStringColumn) {
+  // ARROW-10493
+  auto type =
+      ::arrow::struct_({::arrow::field("inner", ::arrow::utf8(), /*nullable=*/true)});
+  auto outer_array = ::arrow::ArrayFromJSON(type,
+                                            R"([{"inner": "a"},
+                                                {"inner": null},
+                                                {"inner": "b"},
+                                                {"inner": "c"},
+                                                {"inner": null},
+                                                {"inner": null},
+                                                {"inner": "d"},
+                                                null])");
+  auto inner_array =
+      std::static_pointer_cast<::arrow::StructArray>(outer_array)->field(0);
+  ASSERT_OK_AND_ASSIGN(inner_array->data()->buffers[0],
+                       ::arrow::internal::BytesToBits({1, 0, 1, 1, 0, 0, 1, 0}));
+  ASSERT_OK_AND_ASSIGN(outer_array->data()->buffers[0],
+                       ::arrow::internal::BytesToBits({1, 1, 1, 1, 1, 1, 1, 0}));
```

Comment:
I think you mean to assert equality here, but you're overwriting the null bitmaps instead.
```suggestion
  ASSERT_OK_AND_ASSIGN(auto expected_inner_validity,
                       ::arrow::internal::BytesToBits({1, 0, 1, 1, 0, 0, 1, 0}));
  ASSERT_OK_AND_ASSIGN(auto expected_outer_validity,
                       ::arrow::internal::BytesToBits({1, 1, 1, 1, 1, 1, 1, 0}));
  AssertBuffersEqual(*inner_array->data()->buffers[0], *expected_inner_validity);
  AssertBuffersEqual(*outer_array->data()->buffers[0], *expected_outer_validity);
```

Reply:
Thanks, I got confused by the validity of the inner array being 1 when the outer array validity is 0 for the last element. But this is not actually relevant to this issue, so I will just remove this bit of code.